python/ray/tune/suggest/bayesopt.py

```diff
@@ -142,26 +183,56 @@ def register_analysis(self, analysis):
+            self._register_result(params, report)
 
     def on_trial_complete(self, trial_id, result=None, error=False):
-        """Notification for the completion of trial."""
-        if result:
-            self._process_result(trial_id, result)
-        del self._live_trial_mapping[trial_id]
+        """Notification for the completion of trial.
 
-    def _process_result(self, trial_id, result):
-        self.optimizer.register(
-            params=self._live_trial_mapping[trial_id],
-            target=self._metric_op * result[self.metric])
+        Parameters
+        ----------------
+        trial_id (str): Id of the trial.
+            This is a short alphanumerical string.
+        result (dict): Dictionary of result.
+            May be none when some error occurs.
+        error (bool): Boolean representing a previous error state.
+            The result should be None when error is True.
+        """
+        # We try to get the parameters used for this trial
+        params = self._live_trial_mapping.pop(trial_id, None)
+
+        # The results may be None if some exception is raised during the trial.
+        # Also, if the parameters are None (were already processed)
+        # we interrupt the following procedure.
+        # Additionally, if somehow the error is True but
+        # the remaining values are not we also block the method
+        if result is None or params is None or error:
+            return
+
+        # If we don't have to execute some random search steps
+        if len(self._cached_results) >= self.random_search_trials:
+            #  we simply register the obtained result
+            self._register_result(params, result)
+            return
+
+        # We store the results into a temporary cache
+        self._cached_results.append((params, result))
+
+        # If the random search finished,
+        # we update the BO with all the computer points.
+        if len(self._cached_results) == self.random_search_trials:
+            for params, result in self._cached_results:
+                self._register_result(params, result)
+
+    def _register_result(self, params, result):
+        """Register given tuple of params and results."""
+        self.optimizer.register(params, self._metric_op * result[self.metric])
 
     def save(self, checkpoint_dir):
-        trials_object = self.optimizer
-        with open(checkpoint_dir, "wb") as output:
-            pickle.dump(trials_object, output)
+        """Storing current optimizer state."""
+        with open(checkpoint_dir, "wb") as f:
+            pickle.dump(self.optimizer, f)
 
     def restore(self, checkpoint_dir):
-        with open(checkpoint_dir, "rb") as input:
-            trials_object = pickle.load(input)
-        self.optimizer = trials_object
+        """Restoring current optimizer state."""
+        with open(checkpoint_dir, "rb") as f:
+            self.optimizer = pickle.load(f)
```

Comment:
Do we want to save the random search results we've seen?

Reply:
That makes sense, on it!

Reply:
Do you think it should be okay now?